### PR TITLE
Paid Fallback Campaign support

### PR DESCRIPTION
**Pull request by coderberry**

### CODE REVIEW

File: app/controllers/advertisements_controller.rb [hunk 1]
```diff
@@ -260,14 +260,28 @@ def get_fallback_campaign(campaign_relation)
       fallback_campaign_relation = fallback_campaign_relation.where(id: property.assigned_fallback_campaign_ids)
     end
 
-    choose_campaign(fallback_campaign_relation, ignore_budgets: true) || begin
+    # pre-selected fallback
+    chosen_campaign = choose_campaign(fallback_campaign_relation, ignore_budgets: true)
+
+    # paid fallback
+    if rand < ENV.fetch("PAID_FALLBACK_PERCENT", "90").to_f / 100
```
Comment (hopsoft) [inline app/controllers/advertisements_controller.rb]: No need to quote the default of `90`

File: app/controllers/advertisements_controller.rb [hunk 2]
```diff
@@ -260,14 +260,28 @@ def get_fallback_campaign(campaign_relation)
...
+    chosen_campaign ||= begin
       fallback_campaign_relation = campaign_relation.fallback_with_assigned_property_id(property_id)
         .or(campaign_relation.without_assigned_property_ids.fallback_for_property_id(property_id))
       if property.assigned_fallback_campaign_ids.present?
         fallback_campaign_relation = fallback_campaign_relation.where(id: property.assigned_fallback_campaign_ids)
       end
       choose_campaign fallback_campaign_relation, ignore_budgets: true
     end
+
+    chosen_campaign
```
Comment (hopsoft) [inline app/controllers/advertisements_controller.rb]: I'd probably rename `chosen_campaign` to `campaign`.

File: app/javascript/src/app/controllers/select_geo_targets_controller.js [hunk 3]
```diff
@@ -62,6 +62,46 @@ export default class extends Controller {
...
+  collapseAll() {
+    document
+      .querySelectorAll(
+        "[data-target='checkbox-tree-branch.toggle'], [data-target='checkbox-tree-branch.leaves']"
+      )
+      .forEach(elm => {
+        elm.classList.remove('open');
+      });
+  }
+
```
Comment (hopsoft) [inline app/javascript/src/app/controllers/select_geo_targets_controller.js]: Nice

File: db/migrate/20190924203350_add_paid_fallback_to_campaigns.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddPaidFallbackToCampaigns < ActiveRecord::Migration[5.2]
+  def change
+    add_column :campaigns, :paid_fallback, :boolean, default: false
```
Comment (hopsoft) [inline db/migrate/20190924203350_add_paid_fallback_to_campaigns.rb]: I'd put an index on this column to keep ad serving as fast as possible.

File: app/controllers/advertisements_controller.rb (hunk 1 above)
Comment (coderberry) [reply app/controllers/advertisements_controller.rb]: Good call

File: app/controllers/advertisements_controller.rb [hunk 5]
```diff
@@ -261,18 +261,18 @@ def get_fallback_campaign(campaign_relation)
     end
 
     # pre-selected fallback
-    chosen_campaign = choose_campaign(fallback_campaign_relation, ignore_budgets: true)
+    campaign = choose_campaign(fallback_campaign_relation, ignore_budgets: true)
```
Comment (hopsoft) [inline app/controllers/advertisements_controller.rb]: Won't this line prevent the campaign selections below from running?

Comment (coderberry) [reply app/controllers/advertisements_controller.rb]: Oh.. possibly.

Comment (coderberry) [reply app/controllers/advertisements_controller.rb]: Nope. `fallback_campaign_relation` already has the pre-selected scope from above

**Pull request merged**